src/shared/serialize.c

```diff
@@ -480,12 +482,36 @@ int deserialize_pidref(FDSet *fds, const char *value, PidRef *ret) {
 
         e = startswith(value, "@");
         if (e) {
-                int fd = deserialize_fd(fds, e);
+                _cleanup_free_ char *fdstr = NULL, *pidstr = NULL;
+                _cleanup_close_ int fd = -EBADF;
+
+                r = extract_many_words(&e, ":", /* flags = */ 0, &fdstr, &pidstr);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to deserialize pidref '%s': %m", e);
+                if (r == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Cannot deserialize pidref from empty string.");
+
+                assert(r <= 2);
 
+                fd = deserialize_fd(fds, fdstr);
                 if (fd < 0)
                         return fd;
 
-                r = pidref_set_pidfd_consume(ret, fd);
+                /* Before systemd v255, we only serialized the pid and not the pidfd, so make sure we can
+                 * deal with the old serialization format as well. */
```

Comment:
This sounds wrong. The pidfd support is added in 255 after all. Maybe:

"The serialization format is changed after 255.4. In systemd <= 255.4 only pidfd is serialized, but that causes problems when reconstructing pidref (see serialize_pidref for details). After 255.4 pid is serialized too even if we have pidfd, but we still need to support older format."